src/custom_metrics_registry.cc

```diff
@@ -0,0 +1,163 @@
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "custom_metrics_registry.h"
+#include "triton/common/triton_json.h"
+
+namespace triton { namespace backend { namespace python {
+
+MetricRegistry::MetricRegistry(
+    const std::string& labels, TRITONSERVER_MetricFamily* metric_family)
+    : labels_(labels), value_(0.0), metric_(nullptr)
+{
+  std::vector<const TRITONSERVER_Parameter*> labels_params;
+  ParseLabels(labels_params, labels);
+  THROW_IF_TRITON_ERROR(TRITONSERVER_MetricNew(
+      &metric_, metric_family, labels_params.data(), labels_params.size()));
+  for (const auto label : labels_params) {
+    TRITONSERVER_ParameterDelete(const_cast<TRITONSERVER_Parameter*>(label));
+  }
+}
+
+MetricRegistry::~MetricRegistry()
+{
+  if (metric_ != nullptr) {
+    TRITONSERVER_MetricDelete(metric_);
+  }
+}
+
+const std::string&
+MetricRegistry::Labels()
+{
+  return labels_;
+}
+
+void
+MetricRegistry::ParseLabels(
+    std::vector<const TRITONSERVER_Parameter*>& labels_params,
+    const std::string& labels)
+{
+  triton::common::TritonJson::Value labels_json;
+  THROW_IF_TRITON_ERROR(labels_json.Parse(labels));
+
+  std::vector<std::string> members;
+  labels_json.Members(&members);
+  for (const auto& member : members) {
+    std::string value;
+    THROW_IF_TRITON_ERROR(labels_json.MemberAsString(member.c_str(), &value));
+    labels_params.emplace_back(TRITONSERVER_ParameterNew(
+        member.c_str(), TRITONSERVER_PARAMETER_STRING, value.c_str()));
+  }
+}
+
+void
+MetricRegistry::Increment(const double& value)
+{
+  THROW_IF_TRITON_ERROR(TRITONSERVER_MetricIncrement(metric_, value));
+  UpdateValue();
+}
+
+void
+MetricRegistry::SetValue(const double& value)
+{
+  THROW_IF_TRITON_ERROR(TRITONSERVER_MetricSet(metric_, value));
+  UpdateValue();
+}
+
+double
+MetricRegistry::Value()
+{
+  UpdateValue();
+  return value_;
+}
+
+void
+MetricRegistry::UpdateValue()
+{
+  THROW_IF_TRITON_ERROR(TRITONSERVER_MetricValue(metric_, &value_));
+}
+
+MetricFamilyRegistry::MetricFamilyRegistry(
+    const std::string& name, const std::string& description,
+    const MetricKind& kind)
+    : name_(name), description_(description), kind_(kind),
+      metric_family_(nullptr)
+{
+  TRITONSERVER_MetricKind triton_kind = ToTritonServerMetricKind(kind);
+  THROW_IF_TRITON_ERROR(TRITONSERVER_MetricFamilyNew(
+      &metric_family_, triton_kind, name_.c_str(), description_.c_str()));
+}
+
+MetricFamilyRegistry::~MetricFamilyRegistry()
```

Comment:
Just a suggestion, I was wondering do you think it might be better to have a file for `metrics.cc` and `metrics_family.cc`? I don't fully understand the reasoning behind the current breakdown of files. Each of them can have their own `SaveToSharedMemory` and `LoadFromSharedMemory` functionalities.

Reply:
Do you mean to combine the `custom_metrics_registry.h` and `pb_custom_metrics.h` classes to just one metric/metric_family object or do you mean we should just put metric and metric_family into different files? The reason behind having  `custom_metrics_registry.h` and `pb_custom_metrics.h` files is just that I thought it would be clearer to use two different objects instead of using a couple of `#ifdef` in the file.

Reply:
After some thought, I think maybe it would be better to combine both of the objects to just one `metric`/`metric_family`. I was thinking about how to read/write the `TRITONSERVER_MetricFamily`/`TRITONSERVER_Metric` into SHM in both stub and parent processes if we only have one `metric`/`metric_family` object, but it seems like we can just store the memory address like what we do for [`response_factory`](https://github.com/triton-inference-server/python_backend/blob/ad88027900392f7f481b1c9960fc386a32677f08/src/infer_request.h#L137).

Reply:
>Do you mean to combine the custom_metrics_registry.h and pb_custom_metrics.h classes to just one metric/metric_family object or do you mean we should just put metric and metric_family into different files?

I meant to put metric and metric family into two different files. I think combining them to just one file should be fine too.

> Do you mean to combine the custom_metrics_registry.h and pb_custom_metrics.h classes to just one metric/metric_family object or do you mean we should just put metric and metric_family into different files?

Personally, I think the OOP look-and-feel is more important. For example, if I want to look for the metric related changes they would be all in `metric/metric_family.cc`. We do have a lot of ifdefs in other files too.

Reply:
Thanks for the clarification. Rearranged the files.